### ISSUE TITLE
Trouble using the Remotes: url::

### ISSUE THREAD

**Issue by ijlyttle** [r-lib/remotes#288]
Hi,

I am having trouble using this notation in a `DESCRIPTION` file:

```
Imports: 
    pkg.remote
Remotes:
    url::https://github.com/ijlyttle/pkg.remote/archive/master.zip
```

I have a reproducible example; I have tried to be minimal, but it is full of side-effects.

I have three (empty) packages that I have set up on GitHub:

- [pkg.remote](https://github.com/ijlyttle/pkg.remote)
- [pkg.use.github](https://github.com/ijlyttle/pkg.use.github)
- [pkg.use.url](https://github.com/ijlyttle/pkg.use.url)

Both pkg.use.github and pkg.use.url import pkg.remote using the `Remotes` field, but the mechanisms are different:

pkg.use.github:

```
Imports: 
    pkg.remote
Remotes:
    ijlyttle/pkg.remote
```

---

pkg.use.url:

```
Imports: 
    pkg.remote
Remotes:
    url::https://github.com/ijlyttle/pkg.remote/archive/master.zip
```

---

My starting state is a clean R session with *none* of these packages installed.

For pkg.remote, `install_github()` works:

``` r
remotes::install_github("ijlyttle/pkg.remote")
#> Downloading GitHub repo ijlyttle/pkg.remote@master
#>   
   checking for file ‘/private/var/folders/rg/9kvy_lsn43z90d439n02nfw80000gn/T/Rtmp9fRYI7/remotes1091e759d1ad4/ijlyttle-pkg.remote-5926b5b/DESCRIPTION’ ...
  
✔  checking for file ‘/private/var/folders/rg/9kvy_lsn43z90d439n02nfw80000gn/T/Rtmp9fRYI7/remotes1091e759d1ad4/ijlyttle-pkg.remote-5926b5b/DESCRIPTION’
#> 
  
─  preparing ‘pkg.remote’:
#> 
  
   checking DESCRIPTION meta-information ...
  
✔  checking DESCRIPTION meta-information
#> 
  
─  checking for LF line-endings in source and make files and shell scripts
#> 
  
─  checking for empty or unneeded directories
#> ─  building ‘pkg.remote_0.0.0.9000.tar.gz’
#> 
  
   
#> 
remove.packages("pkg.remote")
#> Removing package from '/Library/Frameworks/R.framework/Versions/3.5/Resources/library'
#> (as 'lib' is unspecified)
```

<sup>Created on 2019-01-31 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

For pkg.remote, `install_url()` also works:

``` r
remotes::install_url("https://github.com/ijlyttle/pkg.remote/archive/master.zip")
#> Downloading package from url: https://github.com/ijlyttle/pkg.remote/archive/master.zip
#>   
   checking for file ‘/private/var/folders/rg/9kvy_lsn43z90d439n02nfw80000gn/T/RtmpOHNcOk/remotes10ee13862ad86/pkg.remote-master/DESCRIPTION’ ...
  
✔  checking for file ‘/private/var/folders/rg/9kvy_lsn43z90d439n02nfw80000gn/T/RtmpOHNcOk/remotes10ee13862ad86/pkg.remote-master/DESCRIPTION’
#> 
  
─  preparing ‘pkg.remote’:
#> 
  
   checking DESCRIPTION meta-information ...
  
✔  checking DESCRIPTION meta-information
#> 
  
─  checking for LF line-endings in source and make files and shell scripts
#> 
  
─  checking for empty or unneeded directories
#> ─  building ‘pkg.remote_0.0.0.9000.tar.gz’
#> 
  
   
#> 
remove.packages("pkg.remote")
#> Removing package from '/Library/Frameworks/R.framework/Versions/3.5/Resources/library'
#> (as 'lib' is unspecified)
```

<sup>Created on 2019-01-31 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

I can `install_github()` for pkg.use.github:

``` r

remotes::install_github("ijlyttle/pkg.use.github")
#> Downloading GitHub repo ijlyttle/pkg.use.github@master
#> pkg.remote (NA -> 5926b5bdb...) [GitHub]
#> Downloading GitHub repo ijlyttle/pkg.remote@master
#>   
   checking for file ‘/private/var/folders/rg/9kvy_lsn43z90d439n02nfw80000gn/T/RtmpWwjaQB/remotes1108b28b6fcee/ijlyttle-pkg.remote-5926b5b/DESCRIPTION’ ...
  
✔  checking for file ‘/private/var/folders/rg/9kvy_lsn43z90d439n02nfw80000gn/T/RtmpWwjaQB/remotes1108b28b6fcee/ijlyttle-pkg.remote-5926b5b/DESCRIPTION’
#> 
  
─  preparing ‘pkg.remote’:
#> 
  
✔  checking DESCRIPTION meta-information
#> 
  
─  checking for LF line-endings in source and make files and shell scripts
#> 
  
─  checking for empty or unneeded directories
#> ─  building ‘pkg.remote_0.0.0.9000.tar.gz’
#> 
  
   
#> 
  
   checking for file ‘/private/var/folders/rg/9kvy_lsn43z90d439n02nfw80000gn/T/RtmpWwjaQB/remotes1108b2ae41f8/ijlyttle-pkg.use.github-02638a6/DESCRIPTION’ ...
  
✔  checking for file ‘/private/var/folders/rg/9kvy_lsn43z90d439n02nfw80000gn/T/RtmpWwjaQB/remotes1108b2ae41f8/ijlyttle-pkg.use.github-02638a6/DESCRIPTION’
#> 
  
─  preparing ‘pkg.use.github’:
#> 
  
   checking DESCRIPTION meta-information ...
  
✔  checking DESCRIPTION meta-information
#> 
  
─  checking for LF line-endings in source and make files and shell scripts
#> 
  
─  checking for empty or unneeded directories
#> ─  building ‘pkg.use.github_0.0.0.9000.tar.gz’
#> 
  
   
#> 
remove.packages("pkg.use.github")
#> Removing package from '/Library/Frameworks/R.framework/Versions/3.5/Resources/library'
#> (as 'lib' is unspecified)
remove.packages("pkg.remote")
#> Removing package from '/Library/Frameworks/R.framework/Versions/3.5/Resources/library'
#> (as 'lib' is unspecified)
```

<sup>Created on 2019-01-31 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

But I cannot `install_github()` for pkg.use.url:

``` r
remotes::install_github("ijlyttle/pkg.use.url")
#> Downloading GitHub repo ijlyttle/pkg.use.url@master
#> Skipping 1 packages not available: pkg.remote
#> Installing 1 packages: pkg.remote
#> Error: (converted from warning) package 'pkg.remote' is not available (for R version 3.5.2)
```

<sup>Created on 2019-01-31 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

This might be a clue, I can `install_github()` for pkg.use.url **if** pkg.remote already exists on my system:

``` r
remotes::install_url("https://github.com/ijlyttle/pkg.remote/archive/master.zip")
#> Downloading package from url: https://github.com/ijlyttle/pkg.remote/archive/master.zip
#>   
   checking for file ‘/private/var/folders/rg/9kvy_lsn43z90d439n02nfw80000gn/T/RtmpQOoW98/remotes114b068830f70/pkg.remote-master/DESCRIPTION’ ...
  
✔  checking for file ‘/private/var/folders/rg/9kvy_lsn43z90d439n02nfw80000gn/T/RtmpQOoW98/remotes114b068830f70/pkg.remote-master/DESCRIPTION’
#> 
  
─  preparing ‘pkg.remote’:
#> 
  
   checking DESCRIPTION meta-information ...
  
✔  checking DESCRIPTION meta-information
#> 
  
─  checking for LF line-endings in source and make files and shell scripts
#> 
  
─  checking for empty or unneeded directories
#> ─  building ‘pkg.remote_0.0.0.9000.tar.gz’
#> 
  
   
#> 
remotes::install_github("ijlyttle/pkg.use.url")
#> Downloading GitHub repo ijlyttle/pkg.use.url@master
#> Downloading package from url: https://github.com/ijlyttle/pkg.remote/archive/master.zip
#>   
   checking for file ‘/private/var/folders/rg/9kvy_lsn43z90d439n02nfw80000gn/T/RtmpQOoW98/remotes114b041845efb/pkg.remote-master/DESCRIPTION’ ...
  
✔  checking for file ‘/private/var/folders/rg/9kvy_lsn43z90d439n02nfw80000gn/T/RtmpQOoW98/remotes114b041845efb/pkg.remote-master/DESCRIPTION’
#> 
  
─  preparing ‘pkg.remote’:
#> 
  
   checking DESCRIPTION meta-information ...
  
✔  checking DESCRIPTION meta-information
#> 
  
─  checking for LF line-endings in source and make files and shell scripts
#> 
  
─  checking for empty or unneeded directories
#> ─  building ‘pkg.remote_0.0.0.9000.tar.gz’
#> 
  
   
#> 
  
   checking for file ‘/private/var/folders/rg/9kvy_lsn43z90d439n02nfw80000gn/T/RtmpQOoW98/remotes114b08b6812e/ijlyttle-pkg.use.url-aeee2e8/DESCRIPTION’ ...
  
✔  checking for file ‘/private/var/folders/rg/9kvy_lsn43z90d439n02nfw80000gn/T/RtmpQOoW98/remotes114b08b6812e/ijlyttle-pkg.use.url-aeee2e8/DESCRIPTION’
#> 
  
─  preparing ‘pkg.use.url’:
#> 
  
   checking DESCRIPTION meta-information ...
  
✔  checking DESCRIPTION meta-information
#> 
  
─  checking for LF line-endings in source and make files and shell scripts
#> 
  
─  checking for empty or unneeded directories
#>   
  
─  building ‘pkg.use.url_0.0.0.9000.tar.gz’
#> 
  
   
#> 
remove.packages("pkg.use.url")
#> Removing package from '/Library/Frameworks/R.framework/Versions/3.5/Resources/library'
#> (as 'lib' is unspecified)
remove.packages("pkg.remote")
#> Removing package from '/Library/Frameworks/R.framework/Versions/3.5/Resources/library'
#> (as 'lib' is unspecified)
```

<sup>Created on 2019-01-31 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

I know this has to be one of the longer minimal examples (if it is indeed minimal), so I thank you for your patience to get this far. Also, per usual, I allow for the possibility that I am doing Something Wrong™️.

Thanks!

(P.S. using CRAN version of remotes, 2.0.2)

**Comment by cderv** [comment]
As I was intrigued by this, I did some investigation and I think the source of the issue is that 📦 deps are not parsed correctly with url remote. Or at leas I don't expect this result. Here is a reprex (thks for the dummy 📦 !)

``` r
dir.create(dummy_dir <- tempfile("test-remote"))
url <- "https://github.com/ijlyttle/pkg.use.url/archive/master.zip"
pkg_bundle <- tempfile(tmpdir = dummy_dir, fileext = ".zip")
download.file(url, destfile = pkg_bundle, mode = "wb")
files <- unzip(pkg_bundle, exdir = dummy_dir)
pkg_source <- dirname(files[[1]])
# not sure this is expected
remotes::dev_package_deps(pkg_source)
#> Not on CRAN ----------------------------
#>  package    installed available is_cran remote
#>  pkg.remote NA        NA        TRUE    CRAN


url <- "https://github.com/ijlyttle/pkg.use.github/archive/master.zip"
pkg_bundle <- tempfile(tmpdir = dummy_dir, fileext = ".zip")
download.file(url, destfile = pkg_bundle, mode = "wb")
files <- unzip(pkg_bundle, exdir = dummy_dir)
pkg_source <- dirname(files[[1]])
# with github this is better
remotes::dev_package_deps(pkg_source)
#> Needs update -----------------------------
#>  package    installed available    is_cran remote
#>  pkg.remote NA        5926b5bdb... FALSE   GitHub

unlink(dummy_dir, recursive = TRUE)
```

Not sure how it should work for URL remote. it seems to be parsed correctly, but no name is determined
``` r
pkg <- list(
  package = "pkg.remote",
  remotes = "url::https://github.com/ijlyttle/pkg.remote/archive/master.zip"
)
remotes:::parse_one_remote(pkg$remotes)
#> $url
#> [1] "https://github.com/ijlyttle/pkg.remote/archive/master.zip"
#> 
#> $subdir
#> NULL
#> 
#> attr(,"class")
#> [1] "url_remote" "remote"
remotes:::remote_deps(pkg)
#> Needs update -----------------------------
#>  package installed available is_cran remote
#>  NA      NA        NA        FALSE   URL
```

At the end, I am not sure to see how it should work. But it seems there is something off because remotes expect the dep to be on cran and it is not because it should be a remor URL 🤔 

Hope it helps.

**Comment by ijlyttle** [comment]
I think this gets us closer, thanks @cderv!

**Comment by ijlyttle** [comment]
I noticed this in `install_url.R`:

https://github.com/r-lib/remotes/blob/dedcc067b2b76ff79b69aaace7171a906c6de001/R/install-url.R#L72-L74

I wondered if returning the package name would solve the problem. As a first step, I hard-coded the return value to `"pkg.remote"`. Repeating @cderv's experiment:

``` r
dir.create(dummy_dir <- tempfile("test-remote"))

url <- "https://github.com/ijlyttle/pkg.use.url/archive/master.zip"
pkg_bundle <- tempfile(tmpdir = dummy_dir, fileext = ".zip")
download.file(url, destfile = pkg_bundle, mode = "wb")
files <- unzip(pkg_bundle, exdir = dummy_dir)
pkg_source <- dirname(files[[1]])

# Use hard-coded return value for remote_package_name.url_remote() 
remotes::dev_package_deps(pkg_source)
#> Needs update -----------------------------
#>  package    installed available is_cran remote
#>  pkg.remote NA        NA        FALSE   URL

unlink(dummy_dir, recursive = TRUE)
```

<sup>Created on 2019-02-13 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

🎉: This seems to return what we would expect, although we should also consider `remote_sha.url_remote()`.

Let's try the full installation:

``` r
remotes::install_github("ijlyttle/pkg.use.url")
#> Downloading GitHub repo ijlyttle/pkg.use.url@master
#> pkg.remote (NA -> NA) [URL]
#> Downloading package from url: https://github.com/ijlyttle/pkg.remote/archive/master.zip
#>   
   checking for file ‘/private/var/folders/rg/9kvy_lsn43z90d439n02nfw80000gn/T/RtmpVYyrkw/remotes35b713bfc7c0/pkg.remote-master/DESCRIPTION’ ...
  
✔  checking for file ‘/private/var/folders/rg/9kvy_lsn43z90d439n02nfw80000gn/T/RtmpVYyrkw/remotes35b713bfc7c0/pkg.remote-master/DESCRIPTION’
#> 
  
─  preparing ‘pkg.remote’:
#> 
  
   checking DESCRIPTION meta-information ...
  
✔  checking DESCRIPTION meta-information
#> 
  
─  checking for LF line-endings in source and make files and shell scripts
#> 
  
─  checking for empty or unneeded directories
#> ─  building ‘pkg.remote_0.0.0.9000.tar.gz’
#> 
  
   
#> 
#> Downloading package from url: https://github.com/ijlyttle/pkg.remote/archive/master.zip
#> Skipping pkg.remote, it is already being installed
#>   
   checking for file ‘/private/var/folders/rg/9kvy_lsn43z90d439n02nfw80000gn/T/RtmpVYyrkw/remotes35b76ddfb603/ijlyttle-pkg.use.url-aeee2e8/DESCRIPTION’ ...
  
✔  checking for file ‘/private/var/folders/rg/9kvy_lsn43z90d439n02nfw80000gn/T/RtmpVYyrkw/remotes35b76ddfb603/ijlyttle-pkg.use.url-aeee2e8/DESCRIPTION’
#> 
  
─  preparing ‘pkg.use.url’:
#> 
  
   checking DESCRIPTION meta-information ...
  
✔  checking DESCRIPTION meta-information
#> 
  
─  checking for LF line-endings in source and make files and shell scripts
#> 
  
─  checking for empty or unneeded directories
#> 
  
─  building ‘pkg.use.url_0.0.0.9000.tar.gz’
#> 
  
   
#> 
remove.packages("pkg.remote")
#> Removing package from '/Library/Frameworks/R.framework/Versions/3.5/Resources/library'
#> (as 'lib' is unspecified)
remove.packages("pkg.use.url")
#> Removing package from '/Library/Frameworks/R.framework/Versions/3.5/Resources/library'
#> (as 'lib' is unspecified)
```

<sup>Created on 2019-02-13 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

🎉🎉: This works, too.

Therefore my immediate plan is to write *actual code* for `remote_package_name.url_remote()` and `remote_sha.url_remote()` then to make a PR.

**Comment by ijlyttle** [comment]
Question: As a first idea, could a `remote_url` be an extension of `remote_local`, where the file is downloaded first to a temp directory?

This could be a way just to use the existing code for the methods for `remote_local`.

**Comment by jimhester** [comment]
I think there is just a small bug where we should be unconditionally downloading the remotes if the package name is NA and we are not doing so.

**Comment by ijlyttle** [comment]
😉: https://www.youtube.com/watch?v=LQy1DH38E5g

**Comment by ijlyttle** [comment]
Hi @jimhester,

Could I ask you to expand on your comment?

Would you prefer that I retract #294 and create a new PR along the lines you suggest?

I suspect that's what you are saying, but I'd like to make sure before pulling apart code.

Thanks!

**Comment by jimhester** [comment]
I think this is now fixed, with remotes 2.0.4 I can use `remotes::install_github("ijlyttle/pkg.use.url")` without an error, though the output is a bit confusing due to some of the steps returning NA arguments.

``` r
remotes::install_github("ijlyttle/pkg.use.url")
#> Downloading GitHub repo ijlyttle/pkg.use.url@master
#> NA (NA -> NA) [URL]
#> Skipping 1 packages not available: pkg.remote
#> Downloading package from url: https://github.com/ijlyttle/pkg.remote/archive/master.zip
#> checking for file ‘/private/var/folders/dt/r5s12t392tb5sk181j3gs4zw0000gn/T/RtmpGXB9Qc/remotes10da64957bdf5/pkg.remote-master/DESCRIPTION’ ...
#> ✔  checking for file ‘/private/var/folders/dt/r5s12t392tb5sk181j3gs4zw0000gn/T/RtmpGXB9Qc/remotes10da64957bdf5/pkg.remote-master/DESCRIPTION’
#> ─  preparing ‘pkg.remote’:
#>✔  checking DESCRIPTION meta-information
#> ─  checking for LF line-endings in source and make files and shell scripts
#> ─  checking for empty or unneeded directories
#> ─  building ‘pkg.remote_0.0.0.9000.tar.gz’
#> Installing package into '/Users/jhester/Library/R/3.6/library'
#> (as 'lib' is unspecified)
#> Downloading package from url: https://github.com/ijlyttle/pkg.remote/archive/master.zip
#> Skipping pkg.remote, it is already being installed
#> ✔  checking for file ‘/private/var/folders/dt/r5s12t392tb5sk181j3gs4zw0000gn/T/RtmpGXB9Qc/remotes10da63b0e9549/ijlyttle-pkg.use.url-aeee2e8/DESCRIPTION’
#> ─  preparing ‘pkg.use.url’:
#> ✔  checking DESCRIPTION meta-information
#> ─  checking for LF line-endings in source and make files and shell scripts
#> ─  checking for empty or unneeded directories
#> ─  building ‘pkg.use.url_0.0.0.9000.tar.gz’
#> Installing package into '/Users/jhester/Library/R/3.6/library'
#> (as 'lib' is unspecified)
```

<sup>Created on 2019-06-10 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by ijlyttle** [comment]
Hi @jimhester, just a quick note that I gave this a test using my company's GitHub Enterprise, which was the motivation behind all the issues & PRs, worked fine for me. Thanks again 🎉
